haproxy_exporter.go

```diff
@@ -228,6 +233,11 @@ func NewExporter(uri string, sslVerify bool, selectedServerMetrics map[int]*prom
 		return nil, fmt.Errorf("unsupported scheme: %q", u.Scheme)
 	}
 
+	excludedServerStatesMap := map[string]struct{}{}
```

Comment:
We already have a function filterServerMetrics and pass in a filtered set of metrics to the exporter instance with selectedServerMetrics. What was the motivation to increase the interface of the Exporter type instead of doing the filtering before in the main function? This would also have the advantage of validating the input early similar to what we with the existing metrics filter.

Reply:
The reason for this PR is the situation where you are doing dynamic scaling in haproxy, e.g. when using haproxy-ingress in kubernetes. When you do that, you typically pre-allocate say 100 server slots per backend. Backend servers (pods in kubernetes) will be dynamically added and removed, or change their readiness state. For some backends, nearly all of the server slots will be unused. The server status is dynamic, therefore server metrics must be filtered based on the server status every time the metrics are collected. It cannot be done before that.

Reply:
Apologies, I first misread the variable name and with that misunderstood the behavior. Thanks for the explanation, makes sense!